RAII chapter for idiomatic rust (#2820)
google/comprehensive-rust: src/idiomatic/leveraging-the-type-system/raii/scope_guard.md

# Scope Guards

A scope guard uses the `Drop` trait to ensure cleanup code runs automatically
when a scope exits — even if due to an error.

```rust,editable,compile_fail
use scopeguard::{ScopeGuard, guard};
use std::{
    fs::{self, File},
    io::Write,
};

fn download_successful() -> bool {
    // [...]
    true
}

fn main() {
    let path = "download.tmp";
    let mut file = File::create(path).expect("cannot create temporary file");

    // Set up cleanup immediately after file creation
    let cleanup = guard(path, |path| {
        println!("download failed, deleting: {:?}", path);
        let _ = fs::remove_file(path);
    });

    writeln!(file, "partial data...").unwrap();

    if download_successful() {
        // Download succeeded, keep the file
        let _path = ScopeGuard::into_inner(cleanup);
        println!("Download complete!");
    }
    // Otherwise, the guard runs and deletes the file
}
```

<details>

- This example simulates an HTTP download. We create a temporary file first,
  then use a scope guard to ensure that the file is deleted if the download

>   fails.

Reviewer: ```suggestion
  fails.

- The `scopeguard` crate allows you to conveniently define a single-use `Drop`-based cleanup without defining a custom type with a custom `Drop` implementation.
```

> - The guard is placed directly after creating the file, so even if `writeln!()`

Reviewer: ```suggestion
- The guard is created directly after creating the file, so even if `writeln!()`
```

  fails, the file will still be cleaned up. This ordering is essential for

>   correctness.

Reviewer: ```suggestion
  correctness.

- The `guard()` creates a `ScopeGuard` instance. It a user-defined value (in this case, `path`) and the cleanup closure that later receives this value.
```

- The guard's closure runs on scope exit unless defused with
  `ScopeGuard::into_inner`. In the success path, we defuse it to preserve the

>   file.

Reviewer: ```suggestion
  file.

- A scope guard is similar to the `defer` feature in Go.
```

Reviewer: Not done.

- This pattern is useful when you want fallbacks or cleanup code to run

>   automatically but only if success is not explicitly signaled.

Reviewer: ```suggestion
  automatically but only if success is not explicitly signaled.

- This pattern is also useful when you don't control the cleanup strategy of the resource object. In this example, `File::drop()` closes the file but does not delete it, and we can't change the standard library to delete the instead (nor should we, it is not a good idea anyway).
```

> - This pattern is useful when you want fallbacks or cleanup code to run
>   automatically but only if success is not explicitly signaled.

Reviewer: ```suggestion
- This pattern is ideal for "cleanup on failure" scenarios, where a cleanup should run by default unless a success path is explicitly taken.
```
